mayastor/src/lvs/lvol.rs

```diff
@@ -330,4 +353,59 @@ impl Lvol {
             }
         }
     }
+
+    /// Format snapshot name
+    /// base_name is the nexus or replica UUID
+    pub fn format_snapshot_name(base_name: &str, snapshot_time: u64) -> String {
+        format!("{}-snap-{}", base_name, snapshot_time)
+    }
+
+    /// Create a snapshot
+    pub async fn create_snapshot(
+        self,
+        nvmf_req: *mut spdk_nvmf_request,
+        snapshot_name: &str,
+    ) {
+        extern "C" fn snapshot_done_cb(
+            nvmf_req_ptr: *mut c_void,
+            _lvol_ptr: *mut spdk_lvol,
+            errno: i32,
+        ) {
+            let rsp: &mut spdk_nvme_cpl = unsafe {
+                &mut *spdk_sys::spdk_nvmf_request_get_response(
+                    nvmf_req_ptr as *mut spdk_nvmf_request,
+                )
+            };
+            let nvme_status: &mut spdk_nvme_status =
+                unsafe { &mut rsp.__bindgen_anon_1.status };
+
+            nvme_status.set_sct(0); // SPDK_NVME_SCT_GENERIC
+            nvme_status.set_sc(match errno {
+                0 => 0,
+                _ => {
+                    debug!("vbdev_lvol_create_snapshot errno {}", errno);
+                    0x06 // SPDK_NVME_SC_INTERNAL_DEVICE_ERROR
+                }
+            });
+
+            // From nvmf_bdev_ctrlr_complete_cmd
+            unsafe {
+                spdk_sys::spdk_nvmf_request_complete(
+                    nvmf_req_ptr as *mut spdk_nvmf_request,
```

Comment:
:thinking: we complete the request here... but still create the snapshot?

Reply:
This is part of the callback so the request to create the snapshot will have already been sent.